tmpMod4str = str(tmpMod4)
def get_all_mod4_output_fam_files(wildcards):
    return [f"{tmpMod4str}/{batch}/m4_output_good_markers.fam" for batch in batches]

def get_all_mod4_output_bim_files(wildcards):
    return [f"{tmpMod4str}/{batch}/m4_output_good_markers.bim" for batch in batches]

rule mod5_extend_count_tables:
    input:
        fam_files = get_all_mod4_output_fam_files,
        bim_files = get_all_mod4_output_bim_files,
        previous_samples_table = rules.mod4_extend_count_tables.output.samples_table,
        previous_variants_table = rules.mod4_extend_count_tables.output.variants_table,
        id_mapping = config["id_mapping"]
    output:
        samples_table = resultPath / "count_tables" / "mod5_samples",
        variants_table = resultPath / "count_tables" / "mod5_variants"
    params:
        mod = "mod5"
    conda:
        "envs/r_4.2.yaml"
    shell:
        """
        Rscript utils/extend_count_tables.R \
            {input.id_mapping} \
            {input.previous_samples_table} \
            {output.samples_table} \
            {params.mod} \
            {input.fam_files}
        Rscript utils/extend_count_tables.R \
            {input.id_mapping} \
            {input.previous_variants_table} \
            {output.variants_table} \
            {params.mod} \
            {input.bim_files}
        """

# cull samples for faster debug. bypass by passing number_of_samples = 'all'
# rule_stem = "mod5_cull_samples"
# rule mod5_cull_samples:
#     input:
#         # bedset=multiext("/mnt/work/marc/qc/test_mod5_24.05.02/{batch}/m4_output_good_markers", ".bed", ".bim", ".fam")
#         bedset=rules.m4_output_good_markers.output.m4_output_good_markers
#     output:
#         bedset=multiext(str(tmpMod5 / "{batch}" / "mod5_input_{number_of_samples}_samples"), ".bed", ".bim", ".fam")
#     run:
#         try:
#             inTrunk = mqc.plinkBase(input.bedset[0])
#             outTrunk = mqc.plinkBase(output.bedset[0])
#             if wildcards.number_of_samples.lower() == "all":
#                 subprocess.run(
#                     [
#                         plinklocal,
#                         "--bfile", inTrunk,
#                         "--make-bed",
#                         "--out", outTrunk
#                     ]
#                 )
#             elif wildcards.number_of_samples.isdigit():
#                 samples_txt = f"{outTrunk}.first{wildcards.number_of_samples}samples.txt"
#                 input_bed = input.bedset[2]
#                 cmd = f"head -n {wildcards.number_of_samples} {input_bed}  > {samples_txt}"
#                 subprocess.run(cmd, shell=True, check=True)
#                 subprocess.run(
#                     [
#                         plinklocal,
#                         "--bfile", inTrunk,
#                         "--keep", samples_txt,
#                         "--make-bed",
#                         "--out", outTrunk
#                     ]
#                 )
#             else:
#                 raise Exception("number_of_samples must either be a digit or 'all'")
#         except Exception as e:
#             print(f"An exception occurred in rule {rule}.")
#             print(e)

rule_stem = 'mod5_merge_x'
rule mod5_merge_x:
    input:
        bedset=rules.m4_output_good_markers.output.m4_output_good_markers
    output:
        bedset=multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        inTrunk = mqc.plinkBase(input.bedset[0])
        outTrunk = mqc.plinkBase(output.bedset[0])
        subprocess.run(
            [
                plinklocal,
                "--bfile", inTrunk,
                "--merge-x",
                "--make-bed",
                "--out", outTrunk
            ]
        )


rule_stem = 'mod5_exclude_duplicates'
rule mod5_exclude_duplicates:
    input:
        # bedset=multiext("/mnt/work/marc/qc/test_mod5_24.05.02/{batch}/m4_output_good_markers", ".bed", ".bim", ".fam")
        # bedset=rules.mod4point5_cull_samples.output.bedset # culled bedset for faster debug
        #bedset=rules.mod5_cull_samples.output.bedset # culled bedset for faster debug
        bedset=rules.mod5_merge_x.output.bedset
    output:
        bedset=multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            dupvar = outTrunk + ".dupvar"
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--set-hh-missing",
                    "--list-duplicate-vars", "ids-only", "suppress-first",
                    "--out", outTrunk,
                ]
            )
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--exclude", dupvar,
                    "--make-bed",
                    "--out", outTrunk,
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)


rule_stem = "mod5_exclude_nonstandard_alleles"
rule mod5_exclude_nonstandard_alleles:
    input:
        bedset=rules.mod5_exclude_duplicates.output.bedset
    output:
        bedset=multiext(str(tmpMod5 / "{batch}"  / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--snps-only", "just-acgt",
                    "--make-bed",
                    "--out", outTrunk,
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)

rule_stem = "mod5_exclude_ambiguous_snps"
rule mod5_exclude_ambiguous_snps:
    input:
        bedset = rules.mod5_exclude_nonstandard_alleles.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            mqc.exclude_strand_ambigious_markers(inTrunk, outTrunk, plinklocal)
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)

rule_stem = "mod5_zero_out_mendelian_errors"
rule mod5_zero_out_mendelian_errors:
    input:
        bedset = rules.mod5_exclude_ambiguous_snps.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--me", "1", "1",
                    "--set-me-missing",
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
            # Check if subjects have been removed:
            n_pre = mqc.line_count(inTrunk + ".fam")
            n_post = mqc.line_count(outTrunk + ".fam")
            if n_pre != n_post:
                print("Subjects removed when zeroing out Mendelian errors")
                # should this raise an exception?
                # raise Exception("Subjects removed when zeroing out Mendelian errors")
                
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)


rule_stem_hrc100_input = "mod5_missingness_and_hwe"
rule mod5_missingness_and_hwe:
    input:
        bedset = rules.mod5_zero_out_mendelian_errors.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem_hrc100_input), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--hwe", "1e-6",
                    "--geno", "0.02",
                    "--make-bed",
                    "--out", outTrunk
                ]    
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)


rule mod5_compute_allele_frequencies:
    input:
        bedset = rules.mod5_missingness_and_hwe.output.bedset
    output:
        frq = tmpMod5 / "{batch}" / ("mod5_compute_allele_frequencies.frq")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.frq)
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--freq",
                    #"--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)



rule mod5_extract_Y_mito:
    input:
        bedset = rules.mod5_missingness_and_hwe.output.bedset,
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / "mod5_extract_Y_mito"), ".bed", ".bim", ".fam")
    run:
        inTrunk = mqc.plinkBase(input.bedset[0])
        outTrunk = mqc.plinkBase(output.bedset[0])
        subprocess.run([plinklocal, "--bfile", inTrunk, "--chr", "24", "26", "--make-bed", "--out", outTrunk])

rule_stem_extract_y_mito = "mod5_extract_Y_mito"


# rule_stem must be same as for input bim file
rule mod5_hrc1000:
    input:
        bedset = rules.mod5_missingness_and_hwe.output.bedset,
        frq = rules.mod5_compute_allele_frequencies.output.frq
    output:
        exclude = tmpMod5 / "{batch}"  / ("Exclude-" + rule_stem_hrc100_input + "-HRC.txt"),
        chromosome = tmpMod5 / "{batch}" / ("Chromosome-" + rule_stem_hrc100_input + "-HRC.txt"),
        position = tmpMod5 / "{batch}" / ("Position-" + rule_stem_hrc100_input + "-HRC.txt"),
        strand_flip = tmpMod5 / "{batch}" / ("Strand-Flip-" + rule_stem_hrc100_input + "-HRC.txt"),
        force = tmpMod5 / "{batch}" / ("Force-Allele1-" + rule_stem_hrc100_input + "-HRC.txt"),
        id = tmpMod5 / "{batch}" / ("ID-" + rule_stem_hrc100_input + "-HRC.txt")
    run:
        try:
            subprocess.run(
                [
                    "perl", hrc1000g, 
                    "-b", input.bedset[1],
                    "-f", input.frq,
                    "-r", hrc_sites,
                    "-h"
                ]
            )
            
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)

rule_stem = rule_stem_extract_y_mito
rule mod5_perl_Y_mito:
    input:
        bedset = rules.mod5_extract_Y_mito.output.bedset,
        frq = rules.mod5_compute_allele_frequencies.output.frq
    output:
        exclude = tmpMod5 / "{batch}"  / ("Exclude-" + rule_stem + "-1000G.txt"),
        chromosome = tmpMod5 / "{batch}" / ("Chromosome-" + rule_stem + "-1000G.txt"),
        position = tmpMod5 / "{batch}" / ("Position-" + rule_stem + "-1000G.txt"),
        strand_flip = tmpMod5 / "{batch}" / ("Strand-Flip-" + rule_stem + "-1000G.txt"),
        force = tmpMod5 / "{batch}" / ("Force-Allele1-" + rule_stem + "-1000G.txt"),
        id = tmpMod5 / "{batch}" / ("ID-" + rule_stem + "-1000G.txt")
    run:
        try:
            subprocess.run(
                [
                    "perl", hrc1000g, 
                    "-b", input.bedset[1],
                    "-f", input.frq,
                    "-r", sites_1000g,
                    "-g"
                ]
            )
            
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)

rule_stem = "mod5_exclude"
rule mod5_exclude:
    input:
        exclude = rules.mod5_hrc1000.output.exclude,
        bedset = rules.mod5_missingness_and_hwe.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--exclude", input.exclude,
                    "--make-bed",
                    "--out", outTrunk  
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod5_update_chr"
rule mod5_update_chr:
    input:
        chromosome = rules.mod5_hrc1000.output.chromosome,
        bedset = rules.mod5_exclude.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--update-chr", input.chromosome,
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod5_update_map"
rule mod5_update_map:
    input:
        position = rules.mod5_hrc1000.output.position,
        bedset = rules.mod5_update_chr.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--update-map", input.position,
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod5_strand_flip"
rule mod5_strand_flip:
    input:
        strand_flip = rules.mod5_hrc1000.output.strand_flip,
        bedset = rules.mod5_update_map.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--flip", input.strand_flip,
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod5_force_allele"
rule mod5_force_allele:
    input:
        force = rules.mod5_hrc1000.output.force,
        bedset = rules.mod5_strand_flip.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--a1-allele", input.force,
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod5_update_name"
rule mod5_update_name:
    input:
        id = rules.mod5_hrc1000.output.id,
        bedset = rules.mod5_force_allele.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plinklocal,
                    "--bfile", inTrunk,
                    "--update-name", input.id,
                    "--real-ref-alleles",
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule mod5_rename_missing_ids:
    input:
        # bedset = rules.mod5_remove_blacklisted_variants.output.bedset
        bedset = rules.mod5_update_name.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / "mod5_rename_missing_ids"), ".bed", ".bim", ".fam")
    shell:
        """
        input_full={input.bedset[0]}
        inTrunk="${{input_full%.bed}}"
        output_full={output.bedset[0]}
        outTrunk="${{output_full%.bed}}"
        {plink2local} --bfile $inTrunk --set-missing-var-ids @_#_\$1:\$2 --make-bed --out $outTrunk
        """



# rule_stem = "mod5_remove_variants_with_missing_IDs"
# rule mod5_remove_variants_with_missing_IDs:
#     input:
#         bedset = rules.mod5_update_name.output.bedset
#     output:
#         bedset = multiext(str(tmpMod5 / "{batch}" / "{number_of_samples}_samples" / rule_stem), ".bed", ".bim", ".fam")
#     run:
#         try:
#             inTrunk = mqc.plinkBase(input.bedset[0])
#             outTrunk = mqc.plinkBase(output.bedset[0])
#             subprocess.run(
#                 [
#                     plinklocal,
#                     "--bfile", inTrunk,
#                     "--exclude", exclude_variants,
#                     "--make-bed",
#                     "--out", outTrunk
#                 ]
#             )
#         except Exception as e:
#             print(f"An exception occurred in rule {rule_stem}.")
#             print(e)

rule mod5_split_par:
    input:
        bedset = rules.mod5_rename_missing_ids.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / "mod5_split_par"), ".bed", ".bim", ".fam")
    run:
        inTrunk = mqc.plinkBase(input.bedset[0])
        outTrunk = mqc.plinkBase(output.bedset[0])
        subprocess.run(
            [
                plink2local,
                "--bfile", inTrunk,
                "--split-par", "b37",
                "--make-bed",
                "--out", outTrunk
            ]
        )

rule mod5_set_hh_missing:
    input:
        bedset = rules.mod5_split_par.output.bedset
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / "mod5_set_hh_missing"), ".bed", ".bim", ".fam")
    run:
        inTrunk = mqc.plinkBase(input.bedset[0])
        outTrunk = mqc.plinkBase(output.bedset[0])
        subprocess.run(
            [
                plink2local,
                "--bfile", inTrunk,
                "--set-hh-missing",
                "--make-bed",
                "--out", outTrunk
            ]
        )

rule mod5_sort_vars:
    input:
        bedset = rules.mod5_set_hh_missing.output.bedset
    output:
        pgenset = multiext(str(tmpMod5 / "{batch}" / "mod5_sort_vars"), ".pgen", ".pvar", ".psam")
    run:
        inTrunk = mqc.plinkBase(input.bedset[0])
        outTrunk = mqc.plinkBase(output.pgenset[0])
        subprocess.run(
            [
                plink2local,
                "--bfile", inTrunk,
                "--make-pgen",
                "--sort-vars",
                "--out", outTrunk
            ]
        )

rule_stem = "mod5_output"
rule mod5_output:
    input:
        # bedset = rules.mod5_remove_variants_with_missing_IDs.output.bedset,
        # bedset = rules.mod5_rename_missing_ids.output.bedset,
        #bedset = rules.mod5_set_hh_missing.output.bedset,
        pgenset = rules.mod5_sort_vars.output.pgenset,
        sexcheck = rules.m4_output_good_markers.output.sexcheck
    output:
        bedset = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".bed", ".bim", ".fam"),
        # psam = str(tmpMod5 / "{batch}" / "{number_of_samples}_samples" / (rule_stem + ".psam")),
        missing = multiext(str(tmpMod5 / "{batch}" / rule_stem), ".smiss", ".vmiss"),
        het=str(tmpMod5 / "{batch}" / (rule_stem + ".het")),
        hwe=str(tmpMod5 / "{batch}" /  (rule_stem + ".hardy")),
        md=str(github_docs / "{batch}"/ "pre_imputation_report" / "pre_imputation_report.md")
    run:
        try:
            # Exclude duplicates again
            inTrunk = mqc.plinkBase(input.pgenset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            subprocess.run(
                [
                    plink2local,
                    "--pfile", inTrunk,
                    "--rm-dup", "force-first",
                    "--make-bed",
                    "--out", outTrunk
                ]
            )
            # # keep full original .psam file for later
            # subprocess.run([
            #     plink2local,
            #     "--bfile", outTrunk,
            #     "--make-just-psam",
            #     "--out", outTrunk
            # ])
            subprocess.run([
                plink2local,
                "--bfile", outTrunk,
                "--missing",
                "--het",
                "--hardy",
                "--out", outTrunk
            ])
            batch_report.write_report(output.md, wildcards.batch, "mod5-pre-phasing", outTrunk, input.sexcheck)
        except Exception as e:
            print(f"An exception occurred in rule {rule}.")
            print(e)







