tmpMod5str = str(tmpMod5)
def get_all_mod5_output_fam_files(wildcards):
    return [f"{tmpMod5str}/{batch}/mod5_output.fam" for batch in batches]

def get_all_mod5_output_bim_files(wildcards):
    return [f"{tmpMod5str}/{batch}/mod5_output.bim" for batch in batches]

rule mod6_extend_count_tables:
    input:
        fam_files = get_all_mod5_output_fam_files,
        bim_files = get_all_mod5_output_bim_files,
        previous_samples_table = rules.mod5_extend_count_tables.output.samples_table,
        previous_variants_table = rules.mod5_extend_count_tables.output.variants_table,
        id_mapping = config["id_mapping"]
    output:
        samples_table = resultPath / "count_tables" / "mod6_samples",
        variants_table = resultPath / "count_tables" / "mod6_variants"
    params:
        mod = "mod6"
    conda:
        "envs/r_4.2.yaml"
    shell:
        """
        Rscript utils/extend_count_tables.R \
            {input.id_mapping} \
            {input.previous_samples_table} \
            {output.samples_table} \
            {params.mod} \
            {input.fam_files}
        Rscript utils/extend_count_tables.R \
            {input.id_mapping} \
            {input.previous_variants_table} \
            {output.variants_table} \
            {params.mod} \
            {input.bim_files}
        """

rule mod6_generate_chromosome_file:
    input:
        bim = rules.mod5_output.output.bedset[1]
    output:
        chromosome_file = str(tmpMod6/"{batch}"/"chromosomes.txt")
    run:
        mqc.extract_chromosomes(input.bim, output.chromosome_file)

# tmpMod6str = str(tmpMod6)
# def get_split_output_files(wildcards):
#     chromosomes = mqc.get_chromosome_list(f"{tmpMod6str}/{wildcards.batch}/chromosomes.txt")
#     if wildcards.chr in chromosomes:
#         return [f"{tmpMod6str}/{wildcards.batch}/mod6_split_into_chromosomes.chr{wildcards.chr}.{filetype}" for filetype in [".bed", ".bim", ".fam"]
#     else:
#         return []
    

rule_stem = "mod6_split_into_chromosomes"
rule mod6_split_into_chromosomes:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        bedset = rules.mod5_output.output.bedset
    output:
        bedset = multiext(str(tmpMod6 / "{batch}" / "mod6_split_into_chromosomes.chr{chr}"), ".bed", ".bim", ".fam")
    run:
        try:
            inTrunk = mqc.plinkBase(input.bedset[0])
            outTrunk = mqc.plinkBase(output.bedset[0])
            if mqc.chromosome_in_file(input.chromosome_file, wildcards.chr):
                subprocess.run(
                    [
                        plink2local,
                        "--bfile", inTrunk,
                        "--chr", wildcards.chr,
                        "--make-bed",
                        "--out", outTrunk
                    ]
                )
            else:
                mqc.create_empty_files(output.bedset)
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod6_create_vcf"
rule mod6_create_vcf:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        bedset = rules.mod6_split_into_chromosomes.output.bedset
    output:
        vcf = str(tmpMod6 / "{batch}" / "mod6_create_vcf.chr{chr}.vcf")
    run:
        try:
            if mqc.chromosome_in_file(input.chromosome_file, wildcards.chr):
                inTrunk = mqc.plinkBase(input.bedset[0])
                outTrunk = mqc.plinkBase(output.vcf)
                bim = input.bedset[1]
                cmd = f"{plink2local} --bfile {inTrunk} --recode vcf-iid --ref-allele {bim} 5 2 --out {outTrunk}"
                subprocess.run(cmd, shell = True)
            else:
                mqc.create_empty_files([output.vcf])
        except Exception as e:
            print(f"An exception occurred in rule {rule_stem}.")
            print(e)

rule_stem = "mod6_sort"
rule mod6_sort:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        vcf = rules.mod6_create_vcf.output.vcf
    output:
        sorted_vcf = str(tmpMod6 / "{batch}" / "mod6_sort.chr{chr}.sorted.vcf")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_sort"
            touch {output.sorted_vcf}
        else
            bcftools sort {input.vcf} -o {output.sorted_vcf}
        fi
        """

def get_reference_chr(wildcards):
    if wildcards.chr == "PAR1":
        return "X_PAR1"
    elif wildcards.chr == "PAR2":
        return "X_PAR2"
    elif wildcards.chr == "X":
        return "X_nonPAR"
    else:
        return wildcards.chr

def get_beagle_chr(wildcards):
    if wildcards.chr == "PAR1" or wildcards.chr == "PAR2":
        return "X"
    else:
        return wildcards.chr

rule mod6_conform:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        sorted_vcf = rules.mod6_sort.output.sorted_vcf
    output:
        conformed_vcf = str(tmpMod6 / "{batch}" / "mod6_conform.chr{chr}.conformed.vcf.gz")
    threads: config["mod6conform_threads"]
    params:
        ref_chr=get_reference_chr,
        beagle_chr=get_beagle_chr
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        echo running mod6_conform for chr {wildcards.chr}
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_conform"
            touch {output.conformed_vcf}
        elif [[ {wildcards.chr} == "X" ]]; then
            echo "Bypassing conform for non-PAR X chromosome"
            bgzip -c {input.sorted_vcf} > {output.conformed_vcf}
        else
            reffile=$(echo {hrc_ega}/EGA*/HRC.r1-1.EGA.GRCh37.chr{params.ref_chr}.haplotypes*.vcf.gz)
            output_full={output.conformed_vcf}
            outTrunk="${{output_full%.vcf.gz}}"
            java -Xmx{config[mod6conform_mem]} -jar {conform_gt} gt={input.sorted_vcf} ref=$reffile out=$outTrunk chrom={params.beagle_chr} match=POS
        fi
        """
rule mod6_index_conform:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        vcf = rules.mod6_conform.output.conformed_vcf
    output:
        index = str(tmpMod6 / "{batch}" / "mod6_conform.chr{chr}.conformed.vcf.gz.tbi")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_index_conform"
            touch {output.index}
        else
            tabix -p vcf {input.vcf}
        fi
        """


# Note the fixploidy plugin: male haploid chr X samples must be recoded as diploid homozygotes before phasing due to a bug in shapeit
rule mod6_fill_ac:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        vcf = rules.mod6_conform.output.conformed_vcf,
        index = rules.mod6_index_conform.output.index
    output:
        bcf = str(tmpMod6 / "{batch}" / "mod6_fill_ac.chr{chr}.bcf.gz")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_fill_ac"
            touch {output.bcf}
        else
            bcftools +fill-AN-AC {input.vcf} | bcftools +fixploidy -Ob -o {output.bcf}
        fi
        """


rule mod6_index_fill_ac:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        bcf = rules.mod6_fill_ac.output.bcf
    output:
        index = str(tmpMod6 / "{batch}" / "mod6_fill_ac.chr{chr}.bcf.gz.csi")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_index_fill_ac"
            touch {output.index}
        else
            tabix -p bcf {input.bcf}
        fi
        """


        # shell:
    #     """
    #     reffile=$(echo {hrc_ega}/EGA*/HRC.r1-1.EGA.GRCh37.chr{params.ref_chr}.haplotypes*.vcf.gz)
    #     output_full={output.conformed_vcf}
    #     outTrunk="${{output_full%.vcf.gz}}"
    #     java -Xmx{config[mod6conform_mem]} -jar {conform_gt} gt={input.sorted_vcf} ref=$reffile out=$outTrunk chrom={wildcards.chr} match=POS
    #     """

def get_map_chr(wildcards):
    if wildcards.chr == "PAR1":
        return "X_par1"
    elif wildcards.chr == "PAR2":
        return "X_par2"
    else:
        return wildcards.chr

rule mod6_generate_shapeit_family_files:
    input:
        fam = rules.mod5_output.output.bedset[2]
    output:
        shapeit_fam = str(tmpMod6 / "{batch}" / "shapeit_fam.fam")
    run:
        mqc.filter_fam_table_for_shapeit(input.fam, output.shapeit_fam)

rule mod6_generate_male_list:
    input:
        fam = rules.mod5_output.output.bedset[2]
    output:
        male_list = str(tmpMod6 / "{batch}" / "males.txt"),
        blank_list = str(tmpMod6 / "{batch}" / "blank_list.txt")
    run:
        mqc.create_male_list(input.fam, output.male_list)
        mqc.create_empty_files([output.blank_list])   


tmpMod6str = str(tmpMod6)
def get_shapeit_fam(wildcards):
    return f"{tmpMod6}/{wildcards.batch}/shapeit_fam.fam"

rule mod6_phase:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        shapeit_fam = get_shapeit_fam,
        bcf = rules.mod6_fill_ac.output.bcf,
        index = rules.mod6_index_fill_ac.output.index,
        male_list = rules.mod6_generate_male_list.output.male_list,
        blank_list = rules.mod6_generate_male_list.output.blank_list
    output:
        phased_bcf = str(tmpMod6 / "{batch}" / "mod6_phase.chr{chr}.phased.bcf"),
        shapeit_log = str(tmpMod6 / "{batch}" / "mod6_phase.chr{chr}.phased.log")
    params:
        map_chr=get_map_chr,
        ref_chr=get_reference_chr,
        region_chr=get_beagle_chr
    conda:
        "envs/shapeit_5.yaml"
    threads: 1
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_phase"
            touch {output.phased_bcf}
            touch {output.shapeit_log}
        else
            if [[ "{wildcards.chr}" == "X" ]]; then
                haploid_file={input.male_list}
            else
                haploid_file={input.blank_list}
            fi
            reffile=$(echo {hrc_vcf}/HRC.r1-1.EGA.GRCh37.chr{params.ref_chr}.haplotypes*.vcf.gz)
            mapfile={mapfiles}/shapeit/chr{params.map_chr}.b37.gmap.gz
            SHAPEIT5_phase_common --input {input.bcf} --pedigree {input.shapeit_fam} --region {params.region_chr} --haploids $haploid_file --reference $reffile --map $mapfile --output {output.phased_bcf} --output-format bcf --log {output.shapeit_log} --thread {threads}
        fi
        """


rule mod6_convert_to_vcf:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        bcf = rules.mod6_phase.output.phased_bcf
    output:
        vcf = str(tmpMod6 / "{batch}" / "mod6_phase.chr{chr}.phased.vcf")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_convert_to_vcf"
            touch {output.vcf}
        else
            bcftools convert -Ov -o {output.vcf} {input.bcf}
        fi
        """

rule mod6_generate_bcftools_sex_files:
    input:
        fam = rules.mod5_output.output.bedset[2]
    output:
        sex_list = str(tmpMod6 / "{batch}" / "sex_list.txt")
    run:
        mqc.create_sex_list_for_bcftools(input.fam, output.sex_list) 

# convert male X variants back to haploid before imputation
rule mod6_reset_male_x_ploidy:
    input:
        vcf = rules.mod6_convert_to_vcf.output.vcf,
        sex_list = rules.mod6_generate_bcftools_sex_files.output.sex_list
    output:
        vcf_fixploidy = str(tmpMod6 / "{batch}" / "mod6_reset_male_x_ploidy.chr{chr}.vcf")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        bcftools +fixploidy {input.vcf} -o {output.vcf_fixploidy} -- -p {haploid_x_file} -s {input.sex_list}
        """

rule_stem = "mod6_impute"
rule mod6_impute:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        vcf = rules.mod6_convert_to_vcf.output.vcf #rules.mod6_conform.output.conformed_vcf
    output:
        imputed_vcf = str(tmpMod6 / "{batch}" / "mod6_impute.chr{chr}.imputed.vcf.gz")
    params:
        map_chr=get_map_chr,
        beagle_chr=get_beagle_chr
    threads: config["mod6impute_threads"]
    log:
        str(tmpMod6/"{batch}" / "mod6_impute_log_chr{chr}.log")
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_impute"
            touch {output.imputed_vcf}
        
        else
            reffile={hrc_bref}/HRC_GRCh37.chr{wildcards.chr}.bref3
            mapfile={mapfiles}/plink.chr{params.map_chr}.GRCh37.map
            output_full={output.imputed_vcf}
            outTrunk="${{output_full%.vcf.gz}}"
            java -Xmx{config[mem]} -jar {beagle} gt={input.vcf} ref=$reffile impute=true out=$outTrunk chrom={params.beagle_chr} nthreads={config[beagle_threads]} ap=true map=$mapfile
        fi 
        """

# rule mod6_norm_and_rename:
#     input:
#         imputed_vcf = rules.mod6_impute.output.imputed_vcf
#         # index = rules.mod7_index_batch_merge.output.index
#     output:
#         norm_vcf = str(tmpMod6 / "{batch}" / "{number_of_samples}_samples" / "mod6_norm.chr{chr}.vcf.gz")
#     conda:
#         "envs/bcf_tabix.yaml"
#     shell:
#         """
#         bcftools norm -m + {input.imputed_vcf} | bcftools annotate --set-id +'%CHROM\_%POS\_%REF:%ALT' -o {output.norm_vcf}
#         """
    
rule mod6_index:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        imputed_vcf = rules.mod6_impute.output.imputed_vcf
    output:
        index = str(tmpMod6 / "{batch}" / "mod6_impute.chr{chr}.imputed.vcf.gz.tbi")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_index"
            touch {output.index}
        
        else
            tabix -p vcf {input.imputed_vcf}
        fi 
        
        """

rule mod6_query:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        imputed_vcf = rules.mod6_impute.output.imputed_vcf
    output:
        info = str(tmpMod6 / "{batch}" / "mod6_impute.chr{chr}.imputed.vcf.gz.info")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod6_query"
            touch {output.info}
        
        else
            bcftools query -f '%CHROM %POS %ID %IMP %REF %ALT %DR2 %AF\n' {input.imputed_vcf} -o {output.info}
        fi 
        """






# chrs = chrs_debug # remove in prod!
# tmpMod6str = str(tmpMod6)
# def get_vcf_chr_files_for_batch(wildcards):
#     return [f"{tmpMod6str}/{wildcards.batch}/{wildcards.number_of_samples}_samples/mod6_impute.chr{chr}.imputed.vcf.gz" for chr in chrs]

# def get_index_files_for_batch(wildcards):
#     return [f"{tmpMod6str}/{wildcards.batch}/{wildcards.number_of_samples}_samples/mod6_impute.chr{chr}.imputed.vcf.gz.tbi" for chr in chrs]

# rule mod6_merge_chrs:
#     input:
#         imputed_vcfs = get_vcf_chr_files_for_batch,
#         index = get_index_files_for_batch
#     output:
#         #merged_vcf = str(tmpMod6Archive / "{batch}" / "{number_of_samples}_samples" / "mod6_merge_chrs.vcf.gz")
#         merged_vcf = str(tmpMergeTest / "{batch}" / "{number_of_samples}_samples" / "mod6_merge_chrs.vcf.gz")
#     log:
#         # str(tmpMod6Archive / "{batch}" / "{number_of_samples}_samples" / "mod6_merge.log")
#         str(tmpMergeTest / "{batch}" / "{number_of_samples}_samples" / "mod6_merge.log")
#     conda:
#         "envs/bcf_tabix.yaml"
#     shell:
#         """
#         echo "Output file: {output.merged_vcf}" > {log}
#         echo "Merge started $(date '+%Y-%m-%d %H:%M:%S')" >> {log}
#         bcftools concat {input.imputed_vcfs} -o {output.merged_vcf} --write-index
#         echo "Merge finished $(date '+%Y-%m-%d %H:%M:%S')" >> {log}
#         echo "Merged files:" >> {log}
#         for file in {input.imputed_vcfs}; do echo "$file" >> {log}; done
#         """
# rule mod6_output:
#     input:
#         merged_vcf = rules.mod6_merge_chrs.output.merged_vcf
#     output:
#         # sorted_vcf = str(tmpMod6Archive / "{batch}" / "{number_of_samples}_samples" / "mod6_output.vcf.gz")
#         sorted_vcf = str(tmpMergeTest / "{batch}" / "{number_of_samples}_samples" / "mod6_output.vcf.gz")
#     conda:
#         "envs/bcf_tabix.yaml"
#     shell:
#         """
#         bcftools sort {input.merged_vcf} -o {output.sorted_vcf} --write-index
#         """




# rule mod6_norm:
#     input:
#         imputed_vcf = rules.mod6_impute.output.imputed_vcf
#     output:
#         normalized_vcf = str(tmpMod6 / "{batch}" / "{number_of_samples}_samples" / "mod6_norm.chr{chr}.imputed.vcf.gz")
#     conda:
#         "envs/bcf_tabix.yaml"
#     shell:
#         """
#         bcftools norm -m + {input.imputed_vcf} -o {output.normalized_vcf}
#         """

# rule mod6_convert_to_pgen:
#     input:
#         normalized_vcf = rules.mod6_norm.output.normalized_vcf
#     output:
#         pgenset = multiext(str(tmpMod6 / "{batch}" / "{number_of_samples}_samples" / "mod6_output.chr{chr}"), ".pgen", ".pvar", ".psam")
#     shell:
#         """
#         output_full={output.pgenset[0]}
#         outTrunk="${{output_full%.pgen}}"
#         {plink2local} --vcf {input.normalized_vcf} --make-pgen --out $outTrunk
#         """

# chrs = chrs_debug
# def get_pgen_chr_files_for_batch(wildcards):
#     tmpMod6str = str(tmpMod6)
#     return [f"{tmpMod6str}/{wildcards.batch}/{n_samples}_samples/mod6_output.chr{chr}.pgen" for chr in chrs]
# rule mod6_output:
#     input:
#         pgens = get_pgen_chr_files_for_batch,
#         original_psam = rules.mod5_output.output.psam
#     output:
#         pgenset = multiext(str(tmpMod6 / "{batch}" / "{number_of_samples}_samples" / "mod6_output"), ".pgen", ".pvar", ".psam")
#     run:
#         out_trunk = mqc.plinkBase(output.pgenset[0])
#         # merge chromosomes:
#         mqc.merge_pgensets(input.pgens, out_trunk, plink2local)
#         # add source ID to .psam and restore family information:
#         psam_file = input.original_psam
#         df = pd.read_csv(psam_file, delim_whitespace=True, dtype={'SEX': 'str'})
#         df["SEX"] = df["SEX"].fillna(0)
#         df.insert(2, "SID", wildcards.batch)
#         df.to_csv(output.pgenset[2], sep='\t', index=False)

    