
# NB! REVERT THIS

# rule mod7_filter:
#     input:
#         chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
#         imputed_vcf = rules.mod6_impute.output.imputed_vcf,
#         index = rules.mod6_index.output.index
#     output:
#         filtered_vcf = str(tmpMod6_phasing_test / "dr2_08" / "{batch}" / "mod7_filter.chr{chr}.vcf.gz")
#     conda:
#         "envs/bcf_tabix.yaml"
#     params:
#         dr2_threshold = config['post_imputation_dr2_threshold']
#     shell:
#         """
#         mapfile -t chromosomes < {input.chromosome_file}
#         if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
#             echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod7_filter"
#             touch {output.filtered_vcf}
#         else
#             bcftools view -i 'INFO/DR2>=0.3' {input.imputed_vcf} -o {output.filtered_vcf} --write-index
#         fi 
#         """

def get_dr2_cutoff(wildcards):
    qst = wildcards.quality.split("_")[1]
    qst = f"{qst[0]}.{qst[1]}"
    return float(qst)

def get_imputed_vcf(wildcards):
    return str(tmpMod6_phasing_test / "{wildcards.batch}" / "mod6_impute.chr{wildcards.chr}.imputed.vcf.gz")

def get_imputed_index(wildcards):
    return str(tmpMod6_phasing_test / "{wildcards.batch}" / "mod6_impute.chr{wildcards.chr}.imputed.vcf.gz.tbi")

# def get_chromosome_file(wildcards):
#     return str(tmpMod6/"{wildcards.batch}"/"chromosomes.txt")

rule mod7_filter:
    input:
        # chromosome_file =  rules.mod6_generate_chromosome_file.output.chromosome_file,
        imputed_vcf =  rules.mod6_impute.output.imputed_vcf,
        index = rules.mod6_index.output.index
    output:
        filtered_vcf = str(tmpMod6_phasing_test /"{batch}" / "mod7_filter.chr{chr}.vcf.gz")
    conda:
        "envs/bcf_tabix.yaml"
    # params:
    #     dr2_threshold = get_dr2_cutoff #config['post_imputation_dr2_threshold']
    shell:
        """
        bcftools view -i 'INFO/DR2>=0.3 && INFO/AF>=0.001' {input.imputed_vcf} -o {output.filtered_vcf} --write-index
        """
    
rule mod7_format_phased_dosage:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        filtered_vcf = rules.mod7_filter.output.filtered_vcf
    output:
        formatted_vcf = str(tmpMod7 / "{batch}" / "mod7_format.chr{chr}.vcf.gz")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod7_format_phased_dosage"
            touch {output.formatted_vcf}
        else
            zcat {input.filtered_vcf}  | awk 'BEGIN {{OFS="\\t"}} \
    /^##FORMAT=<ID=(AP1|AP2|DS)/ {{next}} \
    /^##FORMAT=<ID=GT/ {{print; print "##FORMAT=<ID=HDS,Number=.,Type=Float,Description=\\"Phased dosage\\">"; next}} \
    /^#/ {{print; next}} \
    {{split($9, fmt, ":"); \
     if (fmt[1]=="GT" && fmt[2]=="DS" && fmt[3]=="AP1" && fmt[4]=="AP2") {{ \
         $9="GT:HDS"; \
         for (i=10; i<=NF; i++) {{ \
             split($i, vals, ":"); \
             if (length(vals[1]) == 1) {{ \
                 $i=vals[1]":"vals[3]; \
             }} else {{ \
                 $i=vals[1]":"vals[3]","vals[4]; \
             }} \
         }} \
     }} \
     print}}' | bgzip > {output.formatted_vcf}
        fi 
        """




rule mod7_index_format:
    input:
        chromosome_file = rules.mod6_generate_chromosome_file.output.chromosome_file,
        formatted_vcf = rules.mod7_format_phased_dosage.output.formatted_vcf
    output:
        index = str(tmpMod7 / "{batch}" / "mod7_format.chr{chr}.vcf.gz.tbi")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        mapfile -t chromosomes < {input.chromosome_file}
        if [[ ! " ${{chromosomes[@]}} " =~ " {wildcards.chr} " ]]; then
            echo "Chromosome {wildcards.chr} not in chromosome list for {wildcards.batch}, bypassing mod7_index_format"
            touch {output.index}
        else
            tabix -p vcf {input.formatted_vcf}
        fi 
        """

tmpMod6str = str(tmpMod6)
tmpMod7str = str(tmpMod7)

def get_batches_with_chromosome(chr):
    batches_with_chromosome = []
    for batch in batches:
        chromosome_file = f"{tmpMod6str}/{batch}/chromosomes.txt"
        if mqc.chromosome_in_file(chromosome_file, chr):
            batches_with_chromosome.append(batch)
    return batches_with_chromosome

# imputation_batches = ["snp001", "merge_snp002_snp003", "merge_snp007_snp008", "merge_snp009_snp011", "merge_snp010_snp014", "snp012", "snp016a", "snp016b", "snp017a", "merge_snp017b_snp017c_snp017d_snp017e_snp017cf" , "merge_snp018a_snp018b_snp018c_snp018de"]

def get_vcf_batch_files_to_merge(wildcards):
    return [f"{tmpMod6_phasing_test}/{batch}/mod7_filter.chr{wildcards.chr}.vcf.gz" for batch in batches]

    # REVERT THIS!
    # batches_with_chromosome = get_batches_with_chromosome(wildcards.chr)
    # return [f"{tmpMod7str}/{batch}/mod7_format.chr{wildcards.chr}.vcf.gz" for batch in batches_with_chromosome]

tmpMod7str = str(tmpMod7)
def get_index_batch_files_to_merge(wildcards):
    batches_with_chromosome = get_batches_with_chromosome(wildcards.chr)
    return [f"{tmpMod7str}/{batch}/mod7_format.chr{wildcards.chr}.vcf.gz.tbi" for batch in batches_with_chromosome]

rule mod7_merge_batches:
    input:
        vcfs = get_vcf_batch_files_to_merge,
        # index = get_index_batch_files_to_merge # REVERT THIS!
    output:
        merged_vcf = str(tmpMod6_phasing_test / "mod7_merge_batches.chr{chr}.bcf")
    log:
        # str(tmpMod7 / "{number_of_samples}_samples" / "mod7_merge_batches.log")
        str(tmpMod6_phasing_test / "mod7_merge_batches.chr{chr}.log")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        echo "Output file: {output.merged_vcf}" > {log}
        echo "Merge started $(date '+%Y-%m-%d %H:%M:%S')" >> {log}
        bcftools merge {input.vcfs} -o {output.merged_vcf} --write-index
        echo "Merge finished $(date '+%Y-%m-%d %H:%M:%S')" >> {log}
        echo "Merged files:" >> {log}
        for file in {input.vcfs}; do echo "$file" >> {log}; done
        """

rule mod7_index_merge:
    input:
        bcf = rules.mod7_merge_batches.output.merged_vcf
    output:
        index = str(tmpMod6_phasing_test /"mod7_merge_batches.chr{chr}.bcf.csi")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        tabix {input.bcf}
        """

# rule mod7_phase_check:
#     input:
#         bcf = rules.mod7_merge_batches.output.merged_vcf,
#         index = rules.mod7_index_merge.output.index,
#         relations_file = rules.mod6_find_new_batches.output.updated_trios
#     output:
#         phase_check = str(tmpMod6_phasing_test / "mod7_phase_check.{quality}.chr{chr}")
#     conda:
#         "envs/pysam.yaml"
#     params:
#         phase_script = "utils/phasing_check.py"
#     shell:
#         """
#         python {params.phase_script} --trios {input.relations_file} --bcf {input.bcf} --output {output.phase_check} --optimize
#         """


rule mod7_phase_check_parallel:
    input:
        bcf = rules.mod7_merge_batches.output.merged_vcf,
        # index = rules.mod7_index_merge.output.index,
        relations_file = rules.mod6_make_relationships_tables.output.all_relations
    output:
        phase_check = str(tmpMod6_phasing_test / "mod7_phase_check.chr{chr}")
    conda:
        "envs/pysam.yaml"
    params:
        bash_script = "utils/phasing_check_parallel.sh",
        python_script = "utils/phasing_check.py"
    threads: 32
    shell:
        """
        {params.bash_script} {params.python_script} {input.relations_file} {input.bcf} {output.phase_check}  {threads}
        """


rule mod7_phase_report:
    input:
        phase_check = rules.mod7_phase_check_parallel.output.phase_check
    output:
        phase_report = str(tmpMod6_phasing_test / "phase_report" / "mod7_phase_report.{quality}.chr{chr}.md")
    conda:
        "envs/pysam.yaml"
    params:
        r_script = "utils/phasing_report.R",
        title ="\"Phasing report, merged phasing, reshuffled batches\"",
        rate = "phasing_hom",
        rate_desc = "\"Phasing error rates\""
    threads: 32
    conda:
        "envs/r_4.2"
    shell:
        """
        Rscript {params.r_script} {input.phase_check} {output.phase_report} {params.title} {params.rate} {params.rate_desc} {wildcards.chr}
        """


# Plink2 does not support phased dosages for multiallelic variants yet, so they are removed for now
rule mod7_remove_multiallelic_variants:
    input:
        merged_vcf = rules.mod7_merge_batches.output.merged_vcf
        # index = rules.mod7_index_batch_merge.output.index
    output:
        vcf = str(tmpMod7 / "mod7_remove_multiallelic_variants.chr{chr}.vcf.gz")
    conda:
        "envs/bcf_tabix.yaml"
    shell:
        """
        bcftools norm -m + {input.merged_vcf} | bcftools view -m2 -M2 -o {output.vcf}
        """

def get_pre_imputation_fam_files(wildcards):
    tmpMod5str = str(tmpMod5)
    return [f"{tmpMod5str}/{batch}/mod5_output.fam" for batch in batches]

rule mod7_update_sex_file:
    input:
        fams = get_pre_imputation_fam_files
    output:
        sex_file = str(tmpMod7 / "mod7_update_sex_file.txt")
    threads: 24
    run:
        mqc.create_update_sex_file(input.fams, output.sex_file, threads)


rule mod7_convert_chrs_to_pgen:
    input:
        vcf = rules.mod7_remove_multiallelic_variants.output.vcf,
        sex_file = rules.mod7_update_sex_file.output.sex_file
    output:
        pgensets = multiext(str(tmpMod7 / "mod7_convert_chrs_to_pgen.chr{chr}"), ".pgen", ".pvar", ".psam")
    threads: config['mod7_convert_chrs_to_pgen_threads']
    run:
        outTrunk = mqc.plinkBase(output.pgensets[0])
        subprocess.run(
            [
                plink2local,
                "--vcf", input.vcf, "dosage=HDS",
                "--update-sex", input.sex_file,
                "--memory", config['mod7_convert_chrs_to_pgen_mem'],
                "--set-hh-missing", "keep-dosage",
                "--split-par", "b37",
                "--make-pgen",
                "--out", outTrunk
            ]
        )

tmpMod7str = str(tmpMod7)
def chrX_pgens_to_merge(wildcards):
    return [f"{tmpMod7str}/mod7_convert_chrs_to_pgen.chr{chr}.pgen" for chr in ['PAR1', 'X', 'PAR2']]

rule mod7_merge_X:
    input:
        pgens = chrX_pgens_to_merge
    output:
        pgenset = multiext(str(tmpMod7 / "mod7_merge_X"), ".pgen", ".pvar", ".psam")
    run:
        out_trunk = mqc.plinkBase(output.pgenset[0])
        mqc.merge_pgensets(input.pgens, out_trunk, plink2local, 1, pgenlist_file = "merge_x_list")

rule mod7_sort_x:
    input:
        pgenset = rules.mod7_merge_X.output.pgenset
    output:
        pgenset = multiext(str(tmpMod7 / "mod7_sort_X"), ".pgen", ".pvar", ".psam")
    run:
        in_trunk = mqc.plinkBase(input.pgenset[0])
        out_trunk = mqc.plinkBase(output.pgenset[1])
        subprocess.run([
            "plink2",
            "--pfile", in_trunk,
            "--sort-vars",
            "--make-pgen",
            "--out", out_trunk
        ])



def chr_pgens_to_merge(wildcards):
    chrs_to_merge = [f"{tmpMod7str}/mod7_convert_chrs_to_pgen.chr{chr}.pgen" for chr in list([str(i) for i in range(1,23)])]
    chrs_to_merge.append(f"{tmpMod7str}/mod7_sort_X.pgen")
    return chrs_to_merge

rule mod7_merge:
    input:
        pgens = chr_pgens_to_merge
    output:
        pgenset = multiext(str(tmpMod7 / "mod7_merge"), ".pgen", ".pvar", ".psam")
    threads: 32
    run:
        out_trunk = mqc.plinkBase(output.pgenset[0])
        mqc.merge_pgensets(input.pgens, out_trunk, plink2local, threads)


rule mod7_rename_missing_ids:
    input:
        pgenset = rules.mod7_merge.output.pgenset
        #pgenset = multiext("/mnt/work/qc_genotypes/pipeOut_dev/2024.08.29/mod7-post-imputation/all_samples/mod7_output", ".pgen", ".pvar", ".psam")
    output:
        pgenset = multiext(str(tmpMod7 / "mod7_rename_missing_ids"), ".pgen", ".pvar", ".psam")
    shell:
        """
        input_full={input.pgenset[0]}
        inTrunk="${{input_full%.pgen}}"
        output_full={output.pgenset[0]}
        outTrunk="${{output_full%.pgen}}"
        {plink2local} --pfile $inTrunk --set-missing-var-ids @_#_\$1:\$2 --make-pgen --out $outTrunk
        """